tsl/src/remote/connection.c

```diff
@@ -1937,43 +2184,47 @@ remote_connection_cancel_query(TSConnection *conn)
 	 */
 	PG_TRY();
 	{
-		if (conn->status == CONN_COPY_IN && !remote_connection_end_copy(conn, &err))
+		if (conn->status == CONN_COPY_IN && !remote_connection_end_copy_in(conn, &err))
 			remote_connection_error_elog(&err, WARNING);
 
-		/*
-		 * If it takes too long to cancel the query and discard the result, assume
-		 * the connection is dead.
-		 */
-		endtime = TimestampTzPlusMilliseconds(GetCurrentTimestamp(), 30000);
-
 		/*
 		 * Issue cancel request.  Unfortunately, there's no good way to limit the
 		 * amount of time that we might block inside PQcancel().
 		 */
-		if ((cancel = PQgetCancel(conn->pg_conn)))
+		if (!send_cancel(conn))
 		{
-			if (!PQcancel(cancel, errbuf, sizeof(errbuf)))
-			{
-				ereport(WARNING,
-						(errcode(ERRCODE_CONNECTION_FAILURE),
-						 errmsg("could not send cancel request: %s", errbuf)));
-				PQfreeCancel(cancel);
-				remote_connection_set_status(conn, CONN_IDLE);
-				return false;
-			}
-			PQfreeCancel(cancel);
+			remote_connection_set_status(conn, CONN_IDLE);
+
+			if (errmsg != NULL)
+				*errmsg = "cancelation failed";
```

Comment:
Just to be clear that cancel is totally nondeterministic, not sure how much we should care about its successful completion